gtsfm/frontend/verifier/loransac.py

```diff
@@ -0,0 +1,146 @@
+    def __init__(
+        self, use_intrinsics_in_verification: bool, estimation_threshold_px: float, min_allowed_inlier_ratio_est_model: float
+    ) -> None:
+        """Initializes the verifier.
+
+        Args:
+            use_intrinsics_in_verification: Flag to perform keypoint normalization and compute the essential matrix
+                instead of fundamental matrix. This should be preferred when the exact intrinsics are known as opposed
+                to approximating them from exif data.
+            estimation_threshold_px: maximum distance (in pixels) to consider a match an inlier, under squared
+                Sampson distance.
+            min_allowed_inlier_ratio_est_model: minimum allowed inlier ratio w.r.t. the estimated model to accept
+                the verification result and use the image pair, i.e. the lowest allowed ratio of
+                #final RANSAC inliers/ #putatives. A lower fraction indicates less agreement among the result.
+        """
+        self._use_intrinsics_in_verification = use_intrinsics_in_verification
+        self._estimation_threshold_px = estimation_threshold_px
+        self._min_allowed_inlier_ratio_est_model = min_allowed_inlier_ratio_est_model
+        self._min_matches = (
+            NUM_MATCHES_REQ_E_MATRIX
+            if self._use_intrinsics_in_verification
+            else NUM_MATCHES_REQ_F_MATRIX
+        )
+
+        # for failure, i2Ri1 = None, and i2Ui1 = None, and no verified correspondences, and inlier_ratio_est_model = 0
+        self._failure_result = (None, None, np.array([], dtype=np.uint64), 0.0)
+
+    def verify(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        match_indices: np.ndarray,
+        camera_intrinsics_i1: Cal3Bundler,
+        camera_intrinsics_i2: Cal3Bundler,
+    ) -> Tuple[Optional[Rot3], Optional[Unit3], np.ndarray, float]:
+        """Performs verification of correspondences between two images to recover the relative pose and indices of
+        verified correspondences.
+
+        Args:
+            keypoints_i1: detected features in image #i1.
+            keypoints_i2: detected features in image #i2.
+            match_indices: matches as indices of features from both images, of shape (N3, 2), where N3 <= min(N1, N2).
+            camera_intrinsics_i1: intrinsics for image #i1.
+            camera_intrinsics_i2: intrinsics for image #i2.
+
+        Returns:
+            Estimated rotation i2Ri1, or None if it cannot be estimated.
+            Estimated unit translation i2Ui1, or None if it cannot be estimated.
+            Indices of verified correspondences, of shape (N, 2) with N <= N3. These are subset of match_indices.
+            Inlier ratio of w.r.t. the estimated model, i.e. the #final RANSAC inliers/ #putatives.
+        """
+        if match_indices.shape[0] < self._min_matches:
+            logger.info('[LORANSAC] Not enough correspondences for verification.')
```

Comment:
I guess I didn't want to make it an exception because this is how we treat it in other classes.

Getting no matches is possible perhaps on a very difficult scene. Implausible, but still possible i think.

Reply:
Indeed, but that should be handled by the caller. As far this class is concerned, it will fail if there are not enough matches. 
Its already failing, and the better way to do that in python is to raise exceptions rather than returning some dummy result. This way the failure has better "visibility". 
I understand that is not how things are done in other classes, but that should not be a reason to improve (at least in future code). Maybe at some point we can change existing code as well.

Reply:
Yes! Or, you could have the if statement in the calling function. 

if match_indices.shape[0] < threshold:
  log error 
else:
  call function. 
  
As far as this function is concerned, it is an error and we should treat it as one.

Reply:
I added this to https://github.com/borglab/gtsfm/issues/283